primitives/state-machine/src/lib.rs

```diff
@@ -726,6 +731,216 @@ mod execution {
 		prove_read_on_trie_backend(trie_backend, keys)
 	}
 
+	/// State machine only allows a single level
+	/// of child trie.
+	pub const MAX_NESTED_TRIE_DEPTH: usize = 2;
+
+	/// Multiple key value state.
+	/// States are ordered by root storage key.
+	#[derive(PartialEq, Eq, Clone)]
+	pub struct KeyValueStates(pub Vec<KeyValueState>);
+
+	/// A key value state.
+	#[derive(PartialEq, Eq, Clone)]
+	pub struct KeyValueState {
+		/// Storage key in parent states.
+		pub parent_storages: Vec<Vec<u8>>,
```

Comment:
Would it make sense to identify subtree by storage root instead of a key paths?

Reply:
Looks doable: instead of removing root from parent trie on response, we would register their storage key and build a payload that work for client.
It is a bit more work but I think it is a good idea.

Reply:
4bd63f does it, it also avoid adding two child trie with same root in export.